### ISSUE TITLE
Clad Documentation Hackathon.

### ISSUE THREAD

**Issue by parth-07** [vgvassilev/clad#403]
This issue is created to keep track of progress and responsibilities for the Clad documentation hackathon. 

The following concepts along with their person-in-charge are as follows: 
'?' is used if no person is assigned responsibility yet for the particular task.

- [x] Introduction to Clad (Ioana) (most of it can be taken directly from GitHub readme)
- [x] Introduction to automatic differentiation (Ioana) 
- [x] Clad installation and Usage (Ioana) (most of it can be taken directly from GitHub readme)
- [ ] Using Clad
  - [x] Forward mode differentiation (Parth) (most of it can be taken directly from GitHub readme)
  - [ ] Reverse mode differentiation (Baidyanath) (most of it can be taken directly from GitHub readme)
  - [x] Hessian computation (Parth)
  - [x] Jacobian computation (Ioana)
  - [x] Array support (Baidyanath)
  - [x] Functor and lamba support (Parth)
  - [x] Differentiable Class types (Parth)
  - [x] Custom Derivatives (Parth)
- [x] Numerical differentiation fallback (Garima)
- [x] Error Estimation (Garima)
- [ ] Debug Mode (Parth)
- [x] Core Concepts
  - [x] How Clad works (Vassil)
  - [x] The forward and reverse mode in detail (?)
  - [x] Explain how different derived function types are computed (?)
  - [x] Derivative types (Parth)
  - [x] Pushforward and Pullback support (Parth)
  - [x] Custom Derivatives and why we need them. (Parth)
  - [x] Differentiable class type concepts (Parth)
  - [x] Error Estimation core concepts (Garima)
  - [x] Numerical differentiation core concepts (Garima)
- [ ] C++ features support and limitations (?)
- [x] Cuda Support (Baidyanath)
- [x] Clad API
  - [x] `clad::differentiate` (?)
  - [x] `clad::gradient` (?)
  - [x] `clad::hessian` (?)
  - [x] `clad::jacobian` (?)
  - [x] `clad::estimate_error` (Garima)
- [ ] Detailed Demos (all)
- [x] Developer's guide section 
  - [x] Add link to clad internal doxygen documentation (?)
  - [x] Add resources for getting started with Clang frontend transformation and automatic differentiation. (Vassil)
  - [x] Add instructions and tips to set up the clad development environment (?)
- [x] FAQ (?)
- [ ] Improve clad internal doxygen documentation website design (?)
- [x] Add 'CONTRIBUTING.md` file (Vassil)

All of these items should go on the clad readthedocs website. For GitHub readme, we can reuse the introduction and add a link to a few other necessary sections that takes to the readthedocs website. 

This [document](https://docs.google.com/document/d/1BYVZqqYataxGFQVa8Hs0d2jWX26XFBm2pcPUcL4ZLF4/edit?usp=sharing) can be used as a rough outline to follow for the documentation.

**Comment by grimmmyshini** [comment]
Hi Parth! thanks for this issue, can you also just mark what goes in the readme and what in readthedocs? Thanks!

**Comment by vgvassilev** [comment]
I signed up for a bunch of items.

**Comment by grimmmyshini** [comment]
Hi everyone, restructured the readme a little here: https://github.com/grimmmyshini/clad/tree/docs-readme-restructure#clad.

Things left to do in the README:
1. Fill out the "How to contribute" section.
2. Fill out info for clad hessian and jacobian modes.
3. fill out the information in what can be differentiated. (Maybe @parth-07 can add stuff here?)
4. Fill out the links to readthedocs.

**Comment by parth-07** [comment]
> can you also just mark what goes in the readme and what in readthedocs?

All of these tasks should go in the readthedocs. I will mention this in the issue as well. For many items, we can just reuse the existing material from GitHub Readme and already existing material from readthedocs (but it needs to be restructured because we are also changing the layout).

**Comment by parth-07** [comment]
> fill out the information in what can be differentiated. 

I am not sure if we should we keep this section in the GitHub Readme? We want the GitHub readme to be very minimal, right?

**Comment by grimmmyshini** [comment]
> > fill out the information in what can be differentiated.
> 
> I am not sure if we should keep this section in the GitHub Readme. We want the GitHub readme to be very minimal, right?

Yes, but I think we can just add a line there directing to a more elaborate list of what can be differentiated?

**Comment by parth-07** [comment]
> Yes, but I think we can just add a line there directing to a more elaborate list of what can be differentiated?

Yes, this sounds like a good idea.

**Comment by vgvassilev** [comment]
Closing this as split rather than abandoned. The checklist is complete apart from four items, each now its own issue: #2110, #2111, #2112, #2113.

Everything else has landed and was verified against the tree rather than ticked from memory: introduction to AD, installation and usage, forward mode, hessian, jacobian, array support, functors and lambdas, differentiable class types, custom derivatives, numerical differentiation fallback, error estimation, CUDA, all five API entry points, the FAQ, and CONTRIBUTING.md. All nine Core Concepts sub-items are written, including the four that were `.. todo:: todo` placeholders until recently. The developer's-guide items — doxygen link, environment setup, and Clang and AD reading material — are done.

Two things changed while finishing it that are worth recording. Every code example in the user documentation and the README is now compiled and run by check-clad, which found roughly a dozen examples that were wrong, including a README section documenting a custom-derivative API clad had stopped looking up. And docs/userDocs/source/user/SupportedFeatures.rst now answers "C++ features support and limitations", which was the largest remaining box.
